### PR TITLE
Change `PropertyDescriptorGridEntry.NotifyParentsOfChanges` to fix wrong behavior

### LINKED ISSUE #10427: Exception in PropertyGrid when it run in RemoteDesignerHost
### Environment

Microsoft Visual Studio Professional 2022 (64-bit) 17.8.2

### .NET version

.net 6+

### Did this work in a previous version of Visual Studio and/or previous .NET release?

_No response_

### Issue description

Hello. I’m from DevExpress.
We have a problem with PropertyGrid when it run in RemoteDesignerHost:
Screen2

Our research led to the PropertyDescriptorGridEntry class. In the NotifyParentChange method we found a difference in the code for .net 6 and .net framework versions.
Screen3

In .net framework this code look like this
Screen4

I attach an example. In this example in .net version I wrote comment in ComponentChangeService class where an exception occurs in  RemoteDesignerHost class

### Steps to reproduce

1. Download and unzip [Pivot_Error.zip](https://github.com/dotnet/winforms/files/13556605/Pivot_Error.zip)
2. Build and run WinFomNetExample project

### Diagnostics

_No response_

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #10427

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10433)

### CODE REVIEW

**Comment by elachlan** [comment]
> Looks good, had you investigated if it's possible to add unit tests?

It should be possible based on the reproduction project from the issue. I'll look into it.

File: src/System.Windows.Forms/src/System/Windows/Forms/Controls/PropertyGrid/PropertyGridInternal/PropertyDescriptorGridEntry.cs [hunk 1]
```diff
@@ -381,8 +381,8 @@ protected virtual void NotifyParentsOfChanges(GridEntry entry)
             {
                 owner = entry.GetValueOwner();
 
-                ComponentChangeService?.OnComponentChanging(owner, propertyEntry.PropertyDescriptor);
-                ComponentChangeService?.OnComponentChanged(owner, propertyEntry.PropertyDescriptor);
+                ComponentChangeService?.OnComponentChanging(owner, entry.PropertyDescriptor);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Controls/PropertyGrid/PropertyGridInternal/PropertyDescriptorGridEntry.cs:384]: The same fix should be applied to `MultiPropertyDescriptorGridEntry`

**Comment by elachlan** [comment]
@Tanya-Solyanik would this be an integration test? I am unsure how to programmatically set the property value via the `propertygrid` to trigger the behavior.

**Comment by Tanya-Solyanik** [comment]
Would TypeDescriptor.GetProperties + PropertyDescriptor.SetValue work?

**Comment by elachlan** [comment]
I'll give it a go.

**Comment by elachlan** [comment]
Is this basically what you meant?
```csharp
[WinFormsFact]
public void PropertyGrid_NotifyParentChange()
{
    using PropertyGrid propertyGrid = new();
    propertyGrid.Site = new MySite(null, null);
    ErroredCombo combo = new();
    propertyGrid.SelectedObject = combo;
    // combo.Path.ComboName = "123";
    // propertyGrid.Refresh();
    var properties = TypeDescriptor.GetProperties(combo);
    var pathProperty = properties.Find("Path", false);
    var pathObject = pathProperty.GetValue(combo);
    var pathProperties = TypeDescriptor.GetProperties(pathObject);
    var comboNameProperty = pathProperties.Find("ComboName", false);
    comboNameProperty.SetValue(pathObject, "123");
    propertyGrid.Refresh();
}
```

**Comment by elachlan** [comment]
@Tanya-Solyanik Managed to trigger it. The test failed on main and passed with the fixes. Pushed through.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 2]
```diff
@@ -4064,4 +4078,70 @@ public new bool VScroll
 
         public new void WndProc(ref Message m) => base.WndProc(ref m);
     }
+
+    private class ErroredCombo
+    {
+        [TypeConverter(typeof(ExpandableObjectConverter))]
+        public virtual ComboPath Path { get; set; } = new ComboPath() { };
+
+        public class ComboPath
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:4087]:
```suggestion
        public class MyExpandableClass
```
or a better name  :)

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 3]
```diff
@@ -4064,4 +4078,70 @@ public new bool VScroll
...
+            }
+
+            public void OnComponentChanging(object component, MemberDescriptor member)
+            {
+                if (member is null)
+                    return;
+                var props = component.GetType().GetMembers();
+                var name = member.Name;
+                // here exception in RemoteDesignerHost
+                if (props.All(p => p.Name != name))
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:4140]: Assert.False(props.All(p => p.Name != name), "Property not found!");

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 4]
```diff
@@ -4064,4 +4078,70 @@ public new bool VScroll
...
+        {
+            Component = component;
+            Container = container;
+        }
+
+        public IComponent Component { get; private set; }
+
+        public IContainer Container { get; private set; }
+
+        public bool DesignMode => true;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:4110]:
```suggestion
        public bool DesignMode => false;
```

to make this test a little simpler (execute less code)

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 5]
```diff
@@ -4064,4 +4078,70 @@ public new bool VScroll
...
+        [TypeConverter(typeof(ExpandableObjectConverter))]
+        public virtual ComboPath Path { get; set; } = new ComboPath() { };
+
+        public class ComboPath
+        {
+            [NotifyParentProperty(true)]
+            [TypeConverter(typeof(StringConverter))]
+            public string ComboName { get; set; }
+
+            public override string ToString() => $"{{ComboName='{ComboName}'}}";
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:4093]: This override is not needed for the repro

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 6]
```diff
@@ -4064,4 +4078,70 @@ public new bool VScroll
 
         public new void WndProc(ref Message m) => base.WndProc(ref m);
     }
+
+    private class ErroredCombo
+    {
+        [TypeConverter(typeof(ExpandableObjectConverter))]
+        public virtual ComboPath Path { get; set; } = new ComboPath() { };
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:4085]:
```suggestion
        public virtual MyExpandableClass ParenGridEntry { get; set; } = new () { };
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 7]
```diff
@@ -3936,6 +3936,20 @@ public void PropertyGrid_BindObject()
...
+    {
+        using PropertyGrid propertyGrid = new();
+        propertyGrid.Site = new MySite(null, null);
+        ErroredCombo combo = new();
+        propertyGrid.SelectedObject = combo;
+        PropertyGridView propertyGridView = (PropertyGridView)propertyGrid.Controls[2];
+        GridEntry entry = propertyGridView.SelectedGridEntry;
+        PropertyDescriptorGridEntry descriptor = entry.GridItems[0] as PropertyDescriptorGridEntry;
+        descriptor.SetPropertyTextValue("123");
+        propertyGrid.Refresh();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:3950]: I don't think the refresh is needed, we would had failed in the setter, no?

**Review by Tanya-Solyanik — COMMENTED** [review]
Test looks great, I added comments about some simplifications and renaming test classes in terms of winforms mechanics, rather than the old repro business logic.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 8]
```diff
@@ -3936,6 +3936,20 @@ public void PropertyGrid_BindObject()
...
+    public void PropertyGrid_NotifyParentChange()
+    {
+        using PropertyGrid propertyGrid = new();
+        propertyGrid.Site = new MySite(null, null);
+        ErroredCombo combo = new();
+        propertyGrid.SelectedObject = combo;
+        PropertyGridView propertyGridView = (PropertyGridView)propertyGrid.Controls[2];
+        GridEntry entry = propertyGridView.SelectedGridEntry;
+        PropertyDescriptorGridEntry descriptor = entry.GridItems[0] as PropertyDescriptorGridEntry;
+        descriptor.SetPropertyTextValue("123");
```
[the remark this answers is not in the document]
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:3949]: Add a verification that the "inner property" was actually set when you modified the "expandable" property.

Assert.Equal("123", combo.Path.ComboName);

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs [hunk 9]
```diff
@@ -4064,4 +4080,61 @@ public new bool VScroll
...
+            [NotifyParentProperty(true)]
+            [TypeConverter(typeof(StringConverter))]
+            public string NestedGridEntry { get; set; }
+        }
+    }
+
+    private class MySite : ISite
+    {
+        private IComponentChangeService _componentChangeService = new ComponentChangeService();
+        public MySite()
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PropertyGridTests.cs:4100]: Empty default constructor is not needed

**Review by Tanya-Solyanik — APPROVED** [review]

**Pull request merged**